rpc/dial.go

```diff
@@ -278,42 +273,45 @@ func dial(
 	return conn, cached, nil
 }
 
-func dialMulticastDNS(
-	ctx context.Context,
-	address string,
-	logger golog.Logger,
-	dOpts dialOptions,
-) (ClientConn, bool, error) {
+func lookupMDNSCandidate(ctx context.Context, address string, logger golog.Logger) (*zeroconf.ServiceEntry, error) {
 	candidates := []string{address, strings.ReplaceAll(address, ".", "-")}
-	candidateLookup := func(ctx context.Context, candidates []string) (*zeroconf.ServiceEntry, error) {
-		resolver, err := zeroconf.NewResolver(logger, zeroconf.SelectIPRecordType(zeroconf.IPv4))
-		if err != nil {
+	resolver, err := zeroconf.NewResolver(logger, zeroconf.SelectIPRecordType(zeroconf.IPv4))
+	if err != nil {
+		return nil, err
+	}
+	defer resolver.Shutdown()
+	for _, candidate := range candidates {
+		entries := make(chan *zeroconf.ServiceEntry)
+		lookupCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
+		defer cancel()
+		if err := resolver.Lookup(lookupCtx, candidate, "_rpc._tcp", "local.", entries); err != nil {
+			logger.Errorw("error performing mDNS query", "error", err)
 			return nil, err
 		}
-		defer resolver.Shutdown()
-		for _, candidate := range candidates {
-			entries := make(chan *zeroconf.ServiceEntry)
-			lookupCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
-			defer cancel()
-			if err := resolver.Lookup(lookupCtx, candidate, "_rpc._tcp", "local.", entries); err != nil {
-				logger.Errorw("error performing mDNS query", "error", err)
-				return nil, err
-			}
-			select {
-			case <-ctx.Done():
-				return nil, ctx.Err()
-			// entries gets closed after lookupCtx expires or there is a real entry
-			case entry := <-entries:
-				if entry != nil {
-					return entry, nil
-				}
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		// entries gets closed after lookupCtx expires or there is a real entry
+		case entry := <-entries:
+			if entry != nil {
+				return entry, nil
 			}
 		}
+	}
+	if ctx.Err() != nil {
```

Comment:
Did you have a reason to add this extra context check? I don't think it will hurt, but curious if you were consciously thinking about a case where after examining candidates the context has errored?

Reply:
[commented here](https://github.com/viamrobotics/goutils/pull/270/files#r1528461649)

mainly out of caution - it guards against the following scenario:

* select <-entries (meaning an entry is found before context is done)
* entry is nil (L296) suggests this is a possible
* return ctx.Err(), but it's still nil, so we return `nil, nil`

maybe there's some logical reason why this can't happen, but it's not clear to me that it can't happen.

Reply:
Sorry missed your other comment for some reason; cool, it seems fine to include this extra check IMO.